biz.aQute.bndlib/src/aQute/bnd/cdi/CDIAnnotationReader.java

```diff
@@ -0,0 +1,542 @@
+					TypeRef typeRef = analyzer.getTypeRef(classSig.binary);
+					ReferenceDef referenceDef = new ReferenceDef();
+					referenceDef.service = typeRef.getFQN();
+					referenceDef.cardinality = ReferenceCardinality.MULTIPLE;
+					definitions.get(0).references.add(referenceDef);
+					break;
+				default :
+					// skip
+			}
+		}
+	}
+
+	@Override
+	public void parameter(int p) {
+		parameter = p;
+	}
+
+	@Override
+	public void typeuse(int target_type, int target_index, byte[] target_info, byte[] type_path) {
+		if (target_type != 0x10) {
+			targetIndex = -1;
+			return;
+		}
+
+		targetIndex = target_index;
+	}
+
+	private void doReference(Annotation reference, int targetIndex) {
+		Object value = reference.get("value");
+
+		if (value != null) {
+			TypeRef typeRef = (TypeRef) value;
+
+			referenceDef = new ReferenceDef();
+			referenceDef.service = typeRef.getFQN();
+			referenceDef.cardinality = ReferenceCardinality.MANDATORY;
+			definitions.get(0).references.add(referenceDef);
+
+			return;
+		}
+
+		ClassTypeSignature type;
+		ClassResolver resolver;
+		switch (reference.getElementType()) {
+			case PARAMETER : {
+				String signature = (member.getSignature() != null) ? member.getSignature() : member.getDescriptor()
+					.toString();
+				MethodSignature methodSig = analyzer.getMethodSignature(signature);
+				resolver = new MethodResolver(classSig, methodSig);
+				JavaTypeSignature parameterType = ((MethodResolver) resolver).resolveParameter(parameter);
+				if (!(parameterType instanceof ClassTypeSignature)) {
+					analyzer.error("In bean %s, method %s, parameter %s with @Reference has unresolved type %s",
+						classSig, member.getName(), targetIndex, parameterType);
+					return;
+				}
+				type = (ClassTypeSignature) parameterType;
+				break;
+			}
+			default : { // FIELD
+				FieldSignature fieldSig;
+				String signature = member.getSignature();
+				if (signature == null) {
+					try {
+						fieldSig = analyzer.getFieldSignature(member.getDescriptor()
+							.toString());
+					} catch (IllegalArgumentException iae) {
+						fieldSig = null;
+					}
+				} else {
+					fieldSig = analyzer.getFieldSignature(signature);
+				}
+				if (fieldSig == null) {
+					analyzer.error("In bean %s, field %s has an incompatible type for @Reference: %s", clazz,
+						member.getName(), member.getDescriptor());
+					return;
+				}
+				resolver = new FieldResolver(classSig, fieldSig);
+				ReferenceTypeSignature inferred = ((FieldResolver) resolver).resolveField();
+				if (!(inferred instanceof ClassTypeSignature)) {
+					analyzer.error("In bean %s, field %s with @Reference has unresolved type: %s", classSig,
+						member.getName(), inferred);
+					return;
+				}
+				type = (ClassTypeSignature) inferred;
+			}
+		}
+
+		ReferenceCardinality cardinality = ReferenceCardinality.MANDATORY;
+		String fqn = Descriptors.binaryToFQN(type.binary);
+		TypeRef typeRef = analyzer.getTypeRef(type.binary);
```

Comment:
Since you get the TypeRef, it will compute binary, so move the String fqn line below as
`String fqn = typeRef.getFQN();`